modules/blocks/chain.js

```diff
@@ -736,37 +739,49 @@ Chain.prototype.deleteLastBlock = function(cb) {
 		return setImmediate(cb, 'Cannot delete genesis block');
 	}
 
-	async.waterfall(
-		[
-			waterCb => {
-				// Delete last block, replace last block with previous block, undo things
-				__private.popLastBlock(lastBlock, (err, newLastBlock) => {
+	let deletedBlockTransactions;
+
+	async.series(
+		{
+			popLastBlock(seriesCb) {
+				// Perform actual delete of last block
+				__private.popLastBlock(lastBlock, (err, previousBlock) => {
 					if (err) {
 						library.logger.error('Error deleting last block', lastBlock);
+					} else {
+						// Store actual lastBlock transactions in reverse order
+						deletedBlockTransactions = lastBlock.transactions.reverse();
+
+						// Set previous block as our new last block
+						lastBlock = modules.blocks.lastBlock.set(previousBlock);
 					}
-					return setImmediate(waterCb, err, newLastBlock);
+					return setImmediate(seriesCb, err);
 				});
 			},
-			(newLastBlock, waterCb) => {
-				modules.system.update(() => setImmediate(waterCb, null, newLastBlock));
+			updateSystemHeaders(seriesCb) {
+				// Update our own headers: broadhash and height
+				modules.system.update(seriesCb);
 			},
-			(newLastBlock, waterCb) => {
+			broadcastHeaders(seriesCb) {
+				// Notify all remote peers about our new headers
+				modules.transport.broadcastHeaders(seriesCb);
+			},
+			receiveTransactions(seriesCb) {
 				// Put transactions back into transaction pool
 				modules.transactions.receiveTransactions(
-					lastBlock.transactions.reverse(),
+					deletedBlockTransactions,
 					false,
 					err => {
 						if (err) {
 							library.logger.error('Error adding transactions', err);
 						}
-						// Replace last block with previous
-						lastBlock = modules.blocks.lastBlock.set(newLastBlock);
-						return setImmediate(waterCb, null, newLastBlock);
+						deletedBlockTransactions = null;
+						return setImmediate(seriesCb);
```

Comment:
when there is an error in `modules.transactions.receiveTransactions(` the error is not propagated `return setImmediate(seriesCb);`, IMO the error should be propagated `return setImmediate(seriesCb, err);` to have the consistancy of the `Chain.prototype.deleteLastBlock` function.

Reply:
that is correct behavior, error cannot be propagated.